crates/proof-of-sql/src/sql/postprocessing/group_by_postprocessing.rs

```diff
@@ -0,0 +1,322 @@
+            agg_tuples: agg_exprs.into_iter().flatten().collect(),
+        })
+    }
+}
+
+impl<S: Scalar> PostprocessingStep<S> for GroupByPostprocessing {
+    /// Apply the group by transformation to the given `OwnedTable`.
+    fn apply(&self, owned_table: OwnedTable<S>) -> PostprocessingResult<OwnedTable<S>> {
+        dbg!(&self);
+        // First evaluate all the aggregated columns
+        let evaluated_columns: Vec<(AggregationOperator, OwnedColumn<S>, Identifier)> = self
+            .agg_tuples
+            .iter()
+            .map(|(agg_op, expr, id)| {
+                let evaluated_column = owned_table.evaluate(expr)?;
+                Ok((*agg_op, evaluated_column, *id))
+            })
+            .collect::<PostprocessingResult<Vec<_>>>()?;
+        // Next actually do the GROUP BY
+        let alloc = Bump::new();
+        let group_by_columns = self
+            .group_by_identifiers
+            .iter()
+            .map(|id| {
+                let column = owned_table
+                    .inner_table()
+                    .get(id)
+                    .ok_or(PostprocessingError::ColumnNotFound(id.to_string()))?;
+                Ok(Column::<S>::from_owned_column(column, &alloc))
+            })
+            .collect::<PostprocessingResult<Vec<_>>>()?;
+        let (sum_columns, sum_ids, sum_column_types): (Vec<_>, Vec<_>, Vec<_>) = evaluated_columns
+            .iter()
+            .filter(|(agg_op, _owned_col, _id)| *agg_op == AggregationOperator::Sum)
+            .map(|(_agg_op, owned_col, id)| {
+                (
+                    Column::<S>::from_owned_column(owned_col, &alloc),
+                    id,
+                    owned_col.column_type(),
+                )
+            })
+            .multiunzip();
+        let (max_columns, max_ids, max_column_types): (Vec<_>, Vec<_>, Vec<_>) = evaluated_columns
+            .iter()
+            .filter(|(agg_op, _owned_col, _id)| *agg_op == AggregationOperator::Max)
+            .map(|(_agg_op, owned_col, id)| {
+                (
+                    Column::<S>::from_owned_column(owned_col, &alloc),
+                    id,
+                    owned_col.column_type(),
+                )
+            })
+            .multiunzip();
+        let (min_columns, min_ids, min_column_types): (Vec<_>, Vec<_>, Vec<_>) = evaluated_columns
+            .iter()
+            .filter(|(agg_op, _owned_col, _id)| *agg_op == AggregationOperator::Min)
+            .map(|(_agg_op, owned_col, id)| {
+                (
+                    Column::<S>::from_owned_column(owned_col, &alloc),
+                    id,
+                    owned_col.column_type(),
+                )
+            })
+            .multiunzip();
+        let count_ids: Vec<_> = evaluated_columns
+            .iter()
+            .filter(|(agg_op, _owned_col, _id)| *agg_op == AggregationOperator::Count)
+            .map(|(_agg_op, _owned_col, id)| id)
+            .collect::<Vec<_>>();
```

Comment:
Take a look at `Itertools::into_group_map`. It would simplify this code a bunch, and allow you to not collect `evaluated_columns`.